scapy/libs/extcap.py

```diff
@@ -0,0 +1,241 @@
+    res = collections.defaultdict(list)
+    for ifa in data.split("\n"):
+        ifa = ifa.strip()
+        for keyword, values in format.items():
+            if not ifa.startswith(keyword):
+                continue
+
+            def _match(val: str, ifa: str) -> str:
+                m = re.search(r"{%s=([^}]*)}" % val, ifa)
+                if m:
+                    return m.group(1)
+                return ""
+            res[keyword].append(
+                tuple(
+                    [_match(val, ifa) for val in values]
+                )
+            )
+            break
+    return cast(Dict[str, List[Tuple[str, ...]]], res)
+
+
+class _ExtcapNetworkInterface(NetworkInterface):
+    """
+    Extcap NetworkInterface
+    """
+
+    def get_extcap_config(self) -> Dict[str, Tuple[str, ...]]:
+        """
+        Return a list of available configuration options on an extcap interface
+        """
+        return _extcap_call(
+            self.provider.cmdprog,  # type: ignore
+            ["--extcap-interface", self.network_name, "--extcap-config"],
+            {
+                "arg": ["number", "call", "display", "default"],
+                "value": ["arg", "value", "display", "default"],
+            },
+        )
+
+    def get_extcap_cmd(self) -> List[str]:
+        """
+        Return the extcap command line options
+        """
+        cmds = []
+        for x in self.get_extcap_config()["arg"]:
+            if not x[3] or x[3] == "false":
+                continue
+            if x[3] == "true":
+                cmds += [x[1]]
+            else:
+                cmds += [x[3]]
+        return cmds
+
+
+class _ExtcapSocket(SuperSocket):
+    """
+    Read packets at layer 2 using an extcap command
+    """
+
+    nonblocking_socket = True
+
+    @staticmethod
+    def select(sockets: List[SuperSocket],
+               remain: Optional[float] = None) -> List[SuperSocket]:
+        return sockets
+
+    def __init__(self, *_: Any, **kwarg: Any) -> None:
+        cmdprog = kwarg.pop("cmdprog")
+        iface = kwarg.pop("iface", None)
+        if iface is None:
+            raise NameError("Must select an interface for a extcap socket !")
+        iface = resolve_iface(iface)
+        if not isinstance(iface, _ExtcapNetworkInterface):
+            raise ValueError("Interface should be an _ExtcapNetworkInterface")
+        args = iface.get_extcap_cmd()
+        iface = network_name(iface)
+        self.outs = None  # extcap sockets can't write
+        # open fifo
+        fifo, fd = _create_fifo()
+        args = ["--extcap-interface", iface, "--capture", "--fifo", fifo] + args
+        self.proc = subprocess.Popen(
+            [cmdprog] + args,
+        )
+        self.fd = _open_fifo(fd)
+        self.reader = PcapReader(self.fd)  # type: ignore
+        self.ins = self.reader  # type: ignore
+
+    def recv(self, x: int = MTU) -> Packet:
```

Comment:
Shall we raise an exception when using `def send()`?

Reply:
Shouldn't be necessary: l3socket or l2socket can't be built (only l2sniff).